src/lib/limit.ts

```diff
@@ -0,0 +1,164 @@
+import { strict as assert } from 'assert';
+import { promisify } from 'util';
+import { TypedError } from 'typed-error';
+import pThrottle = require('p-throttle');
+
+/**
+ * Generic error thrown when something does not want to be retried/delayed.
+ * This can be things like a backoff or throttle.
+ */
+export class AbortLimit extends TypedError {}
+
+export type ThrottleOptions = {
+	limit: number;
+};
+
+const DEFAULT_THROTTLE_OPTIONS: Partial<ThrottleOptions> = {
+	limit: 1,
+};
+
+/**
+ * Ensures a function is only invoked by default once every N seconds.
+ *
+ * If the func passed throws AbortLimit all queued calls are rejected.
+ * Passing an options with limit greater than 1 means the function can
+ * have parallel calls in progress.
+ */
+export function withThrottle<T extends (...args: any[]) => any>(
+	func: T,
+	wait: number,
+	options: Partial<ThrottleOptions> = {},
+) {
+	assert(typeof func === 'function', 'expected a function as parameter');
+	// If the function returns a promise, unwrap the promise return type
+	// otherwise use the actual return
+	type TReturn = ReturnType<T> extends Promise<infer R> ? R : ReturnType<T>;
+
+	const normalizedOptions: ThrottleOptions = {
+		...DEFAULT_THROTTLE_OPTIONS,
+		...options,
+	} as ThrottleOptions;
+
+	const throttledFunc = pThrottle({
+		limit: normalizedOptions.limit,
+		interval: wait,
+	})(func);
+
+	return async function wrapped(...args: Parameters<T>): Promise<TReturn> {
+		try {
+			return await throttledFunc(...args);
+		} catch (e) {
+			if (e instanceof AbortLimit) {
+				// Remove any queued throttles so the next call happens right away!
+				throttledFunc.abort();
+			}
+			// This will be the exception thrown by func
+			throw e;
+		}
+	};
+}
+
+export type OnRetry = {
```

Comment:
Small nit. I liked the `FailureInfo` name that we had for this type. `OnRetry` is like a name you would use for an event, not for something that is informational.